### PR TITLE
Distribution moments

### PR DESCRIPTION
**Pull request by mjudell**
Re-opening https://github.com/bashtage/arch/pull/328

Added dist.moment(h) for each univariate.Distribution subclass. Also added dist.partial_moment(h, z), which can be useful when working with models that have nonzero asymmetric dependence.  Both methods have closed-form solutions for all distributions.

The test code compares the methods against numbers obtained by numeric integration.

### CODE REVIEW

**Comment by mjudell** [comment]
It looks like the build failure is Travis itself, rather than arch.  Not sure what to do about that. 

Traceback (most recent call last):
  File "/home/travis/miniconda3/envs/arch-test/bin/coveralls", line 5, in <module>
    from coveralls.cli import main
  File "/home/travis/miniconda3/envs/arch-test/lib/python3.6/site-packages/coveralls/__init__.py", line 2, in <module>
    from .api import Coveralls
  File "/home/travis/miniconda3/envs/arch-test/lib/python3.6/site-packages/coveralls/api.py", line 13, in <module>
    from .reporter import CoverallReporter
  File "/home/travis/miniconda3/envs/arch-test/lib/python3.6/site-packages/coveralls/reporter.py", line 10, in <module>
    from coverage.report import Reporter
ImportError: cannot import name 'Reporter'

File: arch/tests/univariate/test_moment.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    (SkewStudent(), [9, 0.1]),
+    (SkewStudent(), [9, 0.5]),
+    (GeneralizedError(), [1.5]),
+    (GeneralizedError(), [2.1]),
+    (StudentsT(), [9]),
+    (StudentsT(), [8]),
+    (Normal(), None)]
+
+
+class MomentTests(object):
```
Comment (bashtage) [inline arch/tests/univariate/test_moment.py]: We should remove the class. It does help here.  Can just have 2 functions.

File: arch/tests/univariate/test_moment.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    """
+    Wrapper around moment and partial moment tests. All it
+    does is ensure the results agree with numeric integrals.
+    For moments 1-5 and partial moments 1-5 evaluated at
+    z= +-1,...,+-5
+    """
+
+    @staticmethod
```
Comment (bashtage) [inline arch/tests/univariate/test_moment.py]: Remove staticmethod once class is gone.

File: arch/tests/univariate/test_moment.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    def test_moments(s, params):
```
Comment (bashtage) [inline arch/tests/univariate/test_moment.py]: Need to parametrize.

```
@pytest.mark.parametrize('dist, params', DISTRIBUTIONS)
def test_moments(dist, params):
```

File: arch/tests/univariate/test_moment.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        if not pmom_good:
+            ptl_moment_failures.append('{:},{:} Partial Moment FAILED'.format(
+                d[0].name, d[1]))
+            all_good = False
+
+        if verbose:
+            print(mom_msg)
+            print(pmom_msg)
+
+    if all_good:
```
Comment (bashtage) [inline arch/tests/univariate/test_moment.py]: No prints please.

File: arch/tests/univariate/test_moment.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            moment_failures.append('{:},{:} Moment FAILED'.format(
+                d[0].name, d[1]))
+            all_good = False
+
+        if not pmom_good:
+            ptl_moment_failures.append('{:},{:} Partial Moment FAILED'.format(
+                d[0].name, d[1]))
+            all_good = False
+
+        if verbose:
```
Comment (bashtage) [inline arch/tests/univariate/test_moment.py]: No verbose.  Should remove this after debugging.

File: arch/tests/univariate/test_moment.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+                all_passed = False
+
+            msg.append(msg_template.format(s.name, params, h, z, cycle_status,
+                                           round(m_formula, 4),
+                                           round(m_integral, 4)))
+
+        return '\n'.join(msg), all_passed
+
+
+def test_moments(verbose=True):
```
Comment (bashtage) [inline arch/tests/univariate/test_moment.py]: Should parmetrize here.

File: arch/tests/univariate/test_moment.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+            rel_err = abs_err / (m_formula + 1e-15)
+
+            if abs_err < eps_abs and (rel_err < eps_rel or
+                                      abs(m_formula) < 1e-8):
+                cycle_status = 'PASSED'
+            else:
+                cycle_status = 'FAILED'
+                all_passed = False
+
+            msg.append(msg_template.format(s.name, params, h, z, cycle_status,
```
Comment (bashtage) [inline arch/tests/univariate/test_moment.py]: Should not use unnecessary strings.

File: arch/tests/univariate/test_moment.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+                else:
+                    m_integral = quad(f, -inf, 0, args=h)[0] + \
+                        quad(f, 0, z, args=h)[0]
+
+            m_formula = s.partial_moment(h, z, params)
+
+            abs_err = abs(m_formula - m_integral)
+            rel_err = abs_err / (m_formula + 1e-15)
+
+            if abs_err < eps_abs and (rel_err < eps_rel or
```
Comment (bashtage) [inline arch/tests/univariate/test_moment.py]: This should be an assert

File: arch/univariate/distribution.py [hunk 9]
```diff
@@ -217,6 +217,57 @@ def cdf(self, resids, parameters=None):
...
+    def moment(self, h, parameters=None):
+        """
+        Moment
+
+        Parameters
+        ----------
+        h : int
+            Order of moment to calculate
+        parameters : ndarray
+            Distribution parameters. Use ''None'' for parameterless
```
Comment (bashtage) [inline arch/univariate/distribution.py]: You can use full width of 80 characters here. Should be 1 line I think.  

Remove '' around None.

File: arch/univariate/distribution.py [hunk 10]
```diff
@@ -217,6 +217,57 @@ def cdf(self, resids, parameters=None):
...
+    @abstractmethod
+    def moment(self, h, parameters=None):
+        """
+        Moment
+
+        Parameters
+        ----------
+        h : int
+            Order of moment to calculate
+        parameters : ndarray
```
Comment (bashtage) [inline arch/univariate/distribution.py]: parameters : ndarray, optional

File: arch/univariate/distribution.py [hunk 11]
```diff
@@ -217,6 +217,57 @@ def cdf(self, resids, parameters=None):
...
+            Distribution parameters. Use ''None'' for parameterless
+            distributions.
+
+        Returns
+        -------
+        moment : float
```
Comment (bashtage) [inline arch/univariate/distribution.py]: Use

```
float
    Calculated moment
```

File: arch/univariate/distribution.py [hunk 12]
```diff
@@ -217,6 +217,57 @@ def cdf(self, resids, parameters=None):
...
+    @abstractmethod
+    def partial_moment(self, h, z=0, parameters=None):
+        r"""
+        Order h partial moment from -inf to z
+
+        Parameters
+        ----------
+        h : int
+            Order of moment
+        z : float
```
Comment (bashtage) [inline arch/univariate/distribution.py]: z : float, optional

File: arch/univariate/distribution.py [hunk 13]
```diff
@@ -217,6 +217,57 @@ def cdf(self, resids, parameters=None):
...
+            Upper bound of partial moment integral
+        parameters : ndarray
```
Comment (bashtage) [inline arch/univariate/distribution.py]: Also needs optional.

File: arch/univariate/distribution.py [hunk 14]
```diff
@@ -217,6 +217,57 @@ def cdf(self, resids, parameters=None):
...
+            Partial moment
+
+        Notes
+        -----
+        The h order partial moment to z is
+
+        .. math::
+
+            \int_{-\infty}^{z}x^{h}f(x)dx
+
```
Comment (bashtage) [inline arch/univariate/distribution.py]: Remove extra line

File: arch/univariate/distribution.py [hunk 15]
```diff
@@ -301,6 +352,26 @@ def ppf(self, pits, parameters=None):
         pits = asarray(pits)
         return stats.norm.ppf(pits)
 
+    def moment(self, h, parameters=None):
+        return stats.norm.moment(h)
+
+    def partial_moment(self, h, z=0, parameters=None):
+        """
```
Comment (bashtage) [inline arch/univariate/distribution.py]: This overwrites previous docstring, so needs full docsting.  Can also use composition to "build" the string and then poke it on.

**Comment by bashtage** [comment]
> Travis is likely using coverage 5.0
> 
> `Traceback (most recent call last): File "/home/travis/miniconda3/envs/arch-test/bin/coveralls", line 5, in <module> from coveralls.cli import main File "/home/travis/miniconda3/envs/arch-test/lib/python3.6/site-packages/coveralls/__init__.py", line 2, in <module> from .api import Coveralls File "/home/travis/miniconda3/envs/arch-test/lib/python3.6/site-packages/coveralls/api.py", line 13, in <module> from .reporter import CoverallReporter File "/home/travis/miniconda3/envs/arch-test/lib/python3.6/site-packages/coveralls/reporter.py", line 10, in <module> from coverage.report import Reporter ImportError: cannot import name 'Reporter'`

I probably need to pin coverage to <5.

**Comment by mjudell** [comment]
Looks like the coverage failure happens outside the additions. I would squash the last 6 commits, but I'm not sure how that would affect the PR.

**Comment by bashtage** [comment]
If you rebase on master you will fix these issues.  If you are familiar, you would need to run the following commands

```bash
git remote add upstream https://github.com/bashtage/arch.git
git fetch --all
git rebase upstream/master
git push
```

**Comment by bashtage** [comment]
Ignore previous.   Yes, slight reductions in coverage happen and aren't actually a blocker.

**Comment by bashtage** [comment]
The distributions.py file has 100% coverage 👍

**Comment by bashtage** [comment]
This has been merged in #333.  Thanks for your contribution.

**Pull request closed without merging**